### ISSUE TITLE
FSDP stuck at "[RANK:0] Starting trainer..."

### ISSUE THREAD

**Issue by RicardoDominguez** [axolotl-ai-cloud/axolotl#1563]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Training using FSDP across several GPUs should work normally.

### Current behaviour

When training using FSDP, the code gets stuck on

```
 [INFO] [axolotl.train.log:61] [PID:1516652] [RANK:0] Starting trainer...
```

After ctrl+c, 

```
[2024-04-23 21:39:51,276] [INFO] [axolotl.train.log:61] [PID:1486799] [RANK:0] Starting trainer...




^C[2024-04-23 22:04:33,011] torch.distributed.elastic.agent.server.api: [WARNING] Received Signals.SIGINT death signal, shutting down workers
[2024-04-23 22:04:33,012] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1486799 closing signal SIGINT
[2024-04-23 22:04:33,012] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1486800 closing signal SIGINT
[2024-04-23 22:04:33,012] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1486801 closing signal SIGINT
[2024-04-23 22:04:33,012] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1486802 closing signal SIGINT
[2024-04-23 22:04:33,012] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1486803 closing signal SIGINT
[2024-04-23 22:04:33,012] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1486804 closing signal SIGINT
[2024-04-23 22:04:33,012] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1486805 closing signal SIGINT
[2024-04-23 22:04:33,012] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1486806 closing signal SIGINT
^C[2024-04-23 22:04:33,183] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1486799 closing signal SIGTERM
[2024-04-23 22:04:33,187] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1486800 closing signal SIGTERM
[2024-04-23 22:04:33,192] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1486801 closing signal SIGTERM
[2024-04-23 22:04:33,194] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1486802 closing signal SIGTERM
[2024-04-23 22:04:33,197] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1486803 closing signal SIGTERM
[2024-04-23 22:04:33,199] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1486804 closing signal SIGTERM
[2024-04-23 22:04:33,201] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1486805 closing signal SIGTERM
[2024-04-23 22:04:33,206] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1486806 closing signal SIGTERM
^CTraceback (most recent call last):
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 736, in run
    result = self._invoke_run(role)
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 877, in _invoke_run
    time.sleep(monitor_interval)
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/api.py", line 62, in _terminate_process_handler
    raise SignalException(f"Process {os.getpid()} got signal: {sigval}", sigval=sigval)
torch.distributed.elastic.multiprocessing.api.SignalException: Process 1486698 got signal: 2

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 743, in run
    self._shutdown(e.sigval)
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/local_elastic_agent.py", line 289, in _shutdown
    self._pcontext.close(death_sig)
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/api.py", line 331, in close
    self._close(death_sig=death_sig, timeout=timeout)
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/api.py", line 713, in _close
    handler.proc.wait(time_to_wait)
  File "/usr/lib/python3.10/subprocess.py", line 1209, in wait
    return self._wait(timeout=timeout)
  File "/usr/lib/python3.10/subprocess.py", line 1953, in _wait
    time.sleep(delay)
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/api.py", line 62, in _terminate_process_handler
    raise SignalException(f"Process {os.getpid()} got signal: {sigval}", sigval=sigval)
torch.distributed.elastic.multiprocessing.api.SignalException: Process 1486698 got signal: 2

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/lustre/home/rolmedo/axo121/bin/accelerate", line 8, in <module>
    sys.exit(main())
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/accelerate/commands/accelerate_cli.py", line 46, in main
    args.func(args)
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/accelerate/commands/launch.py", line 1048, in launch_command
    multi_gpu_launcher(args)
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/accelerate/commands/launch.py", line 702, in multi_gpu_launcher
    distrib_run.run(args)
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/run.py", line 797, in run
    elastic_launch(
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 134, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 255, in launch_agent
    result = agent.run()
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/elastic/metrics/api.py", line 124, in wrapper
    result = f(*args, **kwargs)
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 748, in run
    self._shutdown()
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/local_elastic_agent.py", line 289, in _shutdown
    self._pcontext.close(death_sig)
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/api.py", line 331, in close
    self._close(death_sig=death_sig, timeout=timeout)
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/api.py", line 713, in _close
    handler.proc.wait(time_to_wait)
  File "/usr/lib/python3.10/subprocess.py", line 1209, in wait
    return self._wait(timeout=timeout)
  File "/usr/lib/python3.10/subprocess.py", line 1953, in _wait
    time.sleep(delay)
  File "/lustre/home/rolmedo/axo121/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/api.py", line 62, in _terminate_process_handler
    raise SignalException(f"Process {os.getpid()} got signal: {sigval}", sigval=sigval)
torch.distributed.elastic.multiprocessing.api.SignalException: Process 1486698 got signal: 2
```

Strangely, everything works fine if using qlora.

### Steps to reproduce

With 8 A100 GPUs,

```
accelerate launch -m axolotl.cli.train examples/llama-3/lora-fsdp-70b.yaml
```

### Config yaml

```yaml
base_model: casperhansen/llama-3-70b-fp16
model_type: LlamaForCausalLM
tokenizer_type: AutoTokenizer  # PreTrainedTokenizerFast

load_in_8bit: false
load_in_4bit: true
strict: false

datasets:
  - path: tatsu-lab/alpaca
    type: alpaca
dataset_prepared_path: last_run_prepared
val_set_size: 0.05
output_dir: ./out/qlora-llama3-70b

adapter: lora
lora_model_dir:

sequence_len: 512
sample_packing: false
pad_to_sequence_len: true

lora_r: 8
lora_alpha: 16
lora_dropout: 0.05
lora_target_modules:
lora_target_linear: true
lora_fan_in_fan_out:

wandb_project:
wandb_entity:
wandb_watch:
wandb_name:
wandb_log_model:

gradient_accumulation_steps: 4
micro_batch_size: 1
num_epochs: 4
optimizer: adamw_torch
lr_scheduler: cosine
learning_rate: 0.00001

train_on_inputs: false
group_by_length: false
bf16: auto
fp16:
tf32: false

gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: true
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

warmup_steps: 10
evals_per_epoch: 4
eval_table_size:
saves_per_epoch: 1
debug:
deepspeed:
weight_decay: 0.0
fsdp:
  - full_shard
  - auto_wrap
fsdp_config:
  fsdp_limit_all_gathers: true
  fsdp_sync_module_states: true
  fsdp_offload_params: true
  fsdp_use_orig_params: false
  fsdp_cpu_ram_efficient_loading: true
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_transformer_layer_cls_to_wrap: LlamaDecoderLayer
  fsdp_state_dict_type: FULL_STATE_DICT
  fsdp_sharding_strategy: FULL_SHARD
special_tokens:
  pad_token: <|end_of_text|>
```

### Possible solution

Strangely, everything works fine if using `adapter: qlora`.

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.10.12

### axolotl branch-commit

7d1d22f

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by winglian** [comment]
Load in 8bit instead of 4bit for lora perhaps?

**Comment by RicardoDominguez** [comment]
Tried 8bit and 16bit, same issue.

```
base_model: /fast/rolmedo/models/llama-3-70b/snapshots/model/
model_type: LlamaForCausalLM
tokenizer_type: AutoTokenizer  # PreTrainedTokenizerFast

load_in_8bit: false
load_in_4bit: false
strict: false

datasets:
  - path: tatsu-lab/alpaca
    type: alpaca
dataset_prepared_path: last_run_prepared
val_set_size: 0.05
output_dir: ./out/qlora-llama3-70b

adapter: lora
lora_model_dir:

sequence_len: 512
sample_packing: false
pad_to_sequence_len: true

lora_r: 8
lora_alpha: 16
lora_dropout: 0.05
lora_target_modules:
lora_target_linear: true
lora_fan_in_fan_out:

wandb_project: mistral-lawma
wandb_entity:
wandb_watch:
wandb_name:
wandb_log_model:

gradient_accumulation_steps: 4
micro_batch_size: 1
num_epochs: 4
optimizer: adamw_torch
lr_scheduler: cosine
learning_rate: 0.00001

train_on_inputs: false
group_by_length: false
bf16: true
fp16: false
tf32: false

gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: true
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

warmup_steps: 10
evals_per_epoch: 4
eval_table_size:
saves_per_epoch: 1
debug:
deepspeed:
weight_decay: 0.0
fsdp:
  - full_shard
  - auto_wrap
fsdp_config:
  fsdp_limit_all_gathers: true
  fsdp_sync_module_states: true
  fsdp_offload_params: true
  fsdp_use_orig_params: false
  fsdp_cpu_ram_efficient_loading: true
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_transformer_layer_cls_to_wrap: LlamaDecoderLayer
  fsdp_state_dict_type: FULL_STATE_DICT
  fsdp_sharding_strategy: FULL_SHARD
special_tokens:
  pad_token: <|end_of_text|>
```

**Comment by RicardoDominguez** [comment]
Tried `adapter: ` (for FFT) without luck, as well as different nodes. It is only working with qlora, which is very strange.

**Comment by RicardoDominguez** [comment]
When using lora on smaller models on a single GPU, everything works fine.

**Comment by winglian** [comment]
My guess is it's some combination of the various FSDP options that is causing the issue

**Comment by ryj0902** [comment]
I experienced the same issue.  
This occurred when I copied the qlora example for full fine tuning `Meta-Llama-3-8B-Instruct` using two A100 GPUs.  
As @winglian advised, there seems to be a problem with the FSDP option, but I'm not sure which option exactly caused the problem.  
I reduced the options to a minimum(mentioned [here](https://github.com/OpenAccess-AI-Collective/axolotl?tab=readme-ov-file#fsdp)) and the problem went away:
```
fsdp:
  - full_shard
  - auto_wrap
fsdp_config:
  fsdp_offload_params: true
  fsdp_state_dict_type: FULL_STATE_DICT
  fsdp_transformer_layer_cls_to_wrap: LlamaDecoderLayer
special_tokens:
```
